python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,726 @@
+                    reused_nodes_dict[vm_id] = vm
+                    # Tag needs vm.vm and not vm.name as id
+                    reuse_node_ids.append(vm.vm)
+                    number_of_reused_nodes += 1
+
+            if reuse_nodes:
+                cli_logger.info(
+                    "Reusing nodes {}. "
+                    "To disable reuse, set `cache_stopped_nodes: False` "
+                    "under `provider` in the cluster configuration.",
+                    cli_logger.render_list(reuse_node_ids),
+                )
+                for node_id in reuse_node_ids:
+                    cli_logger.info("Powering on VM")
+                    self.vsphere_sdk_client.vcenter.vm.Power.start(node_id)
+                counter -= len(reuse_node_ids)
+
+        created_nodes_dict = {}
+        if counter:
+            created_nodes_dict = self._create_node(node_config, filters, counter)
+
+        all_created_nodes = reused_nodes_dict
+        all_created_nodes.update(created_nodes_dict)
+
+        # Set tags on the nodes that were reused
+        for node_id in reuse_node_ids:
+            self.set_node_tags(node_id, filters)
+
+        return all_created_nodes
+
+    def get_vm(self, node_id):
+        """
+        Return the VM summary object
+        Note: The method assumes that there is only one vm with the mentioned name.
+        """
+        vm = self._get_cached_node(node_id)
+        cli_logger.info(f"VM {node_id} found")
+
+        return vm
+
+    def attach_tag(self, vm_id, resource_type, tag_id):
+        dyn_id = DynamicID(type=resource_type, id=vm_id)
+        try:
+            self.vsphere_sdk_client.tagging.TagAssociation.attach(tag_id, dyn_id)
+            cli_logger.info("Tag attached")
+        except Exception as e:
+            cli_logger.info(
+                "Check that the tag is associable to {}".format(resource_type)
+            )
+            raise e
+
+    def set_cloudinit_userdata(self, vm_id):
+        logger.info("Setting cloudinit userdata for vm {}".format(vm_id))
+
+        metadata = '{"cloud_name": "vSphere"}'
+
+        # Read the public key that was generated previously.
+        with open(PUBLIC_KEY_PATH, "r") as file:
+            public_key = file.read().rstrip("\n")
+
+        # This file contains the userdata with default values.
+        # We want to add the key that we generate with create_key_pair
+        # function into authorized_keys section
+
+        f = open(USER_DATA_FILE_PATH, "r")
+        data = yaml.load(f, Loader=yaml.FullLoader)
+        for _, v in data.items():
+            for user in v:
+                if isinstance(user, dict):
+                    user["ssh_authorized_keys"] = [public_key]
+        f.close()
+
+        modified_userdata = yaml.dump(data, default_flow_style=False)
+
+        # The userdata needs to be prefixed with #cloud-config.
+        # Without it, the cloudinit spec would get applied but
+        # it wouldn't add the userdata on the VM.
```

Comment:
Can we link to some documentation for this #cloud-config requirement, if it exists?

Reply:
Done